### PR TITLE
K3s/RKE2 Provisioning CI Improvments for Performance/Reliability

### PR DESCRIPTION
**Pull request by Oats87**
Working on: https://github.com/rancher/rancher/pull/36134

Part of https://github.com/rancher/rancher/issues/36162

### CODE REVIEW

**Comment by Oats87** [comment]
```
[provisioning-tests:160] === RUN   TestSystemAgentVersion
[provisioning-tests:161] time="2022-01-14T07:46:17Z" level=info msg="Running in single server mode, will not peer connections"
[provisioning-tests:162] --- PASS: TestSystemAgentVersion (0.03s)
[provisioning-tests:163] === RUN   TestCustomOneNode
[provisioning-tests:164] time="2022-01-14T07:46:18Z" level=info msg="Running in single server mode, will not peer connections"
[provisioning-tests:165] --- PASS: TestCustomOneNode (239.55s)
[provisioning-tests:166] === RUN   TestCustomThreeNode
[provisioning-tests:167] time="2022-01-14T07:50:17Z" level=info msg="Running in single server mode, will not peer connections"
[provisioning-tests:168] --- PASS: TestCustomThreeNode (608.45s)
[provisioning-tests:169] === RUN   TestCustomUniqueRoles
[provisioning-tests:170] time="2022-01-14T08:00:25Z" level=info msg="Running in single server mode, will not peer connections"
[provisioning-tests:171] --- PASS: TestCustomUniqueRoles (775.46s)
[provisioning-tests:172] === RUN   TestCustomThreeNodeWithTaints
```
the tests definitely take a while

File: .drone.yml [hunk 1]
```diff
@@ -48,9 +48,6 @@ steps:
   - name: docker
     path: /var/run/docker.sock
   when:
-    instance:
-      include:
-      - drone-publish.rancher.io
```
Comment (thedadams) [inline .drone.yml:53]: Are we absolutely sure we want to do this?

File: scripts/provisioning-tests [hunk 2]
```diff
@@ -1,28 +1,80 @@
...
+    if [ $PID != -1 ]; then
+      kill $PID
+      wait $PID
+    fi
+    return $EXIT
+}
 
 cd $(dirname $0)/..
 
+SOME_K8S_VERSION=v1.21.5+rke2r2
```
Comment (thedadams) [inline scripts/provisioning-tests:20]: Can we put a comment here indicating that this should be removed when a proper version of RKE2 is added to KDM to address the the problem?

File: tests/integration/pkg/systemdnode/systemnode.go [hunk 3]
```diff
@@ -22,21 +25,62 @@ func New(clients *clients.Clients, namespace, script string) (*corev1.Pod, error
...
+					},
+				},
+				{
+					Name: "agent-images",
+					VolumeSource: corev1.VolumeSource{
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+			InitContainers: []corev1.Container{{
```
Comment (thedadams) [inline tests/integration/pkg/systemdnode/systemnode.go:61]: Why is this necessary? Why can't we just mount the `/image-assets` directory into the container (not the init container)?

**Review by thedadams — CHANGES_REQUESTED** [review]

File: scripts/provisioning-tests [hunk 4]
```diff
@@ -1,28 +1,80 @@
...
-./scripts/run >/tmp/rancher.log 2>&1 &
-PID=$!
 
-keepalive()
+mkdir -p /image-assets
+curl -sLf https://github.com/$TB_ORG/$DIST/releases/download/$SOME_K8S_VERSION/$DIST$AIRGAP-images$LINUX-amd64.tar.gz -o /image-assets/$DIST$AIRGAP-images$LINUX-amd64.tar.gz
+echo Decompressing airgap tarball
+gzip -d /image-assets/$DIST$AIRGAP-images$LINUX-amd64.tar.gz
+
+run_rancher()
```
Comment (thedadams) [inline scripts/provisioning-tests:54]: This function is causing a lot of logs in drone. Can we quiet this down?

File: tests/integration/pkg/systemdnode/systemnode.go (hunk 3 above)
Comment (Oats87) [reply tests/integration/pkg/systemdnode/systemnode.go:61]: The reason is due to the fact that it appears that RKE2 actually mutates the directory to indicate it wants to pull other files

File: scripts/provisioning-tests [hunk 5]
```diff
@@ -1,28 +1,80 @@
...
+  LINUX=.linux
+fi
+
+export DIST=${DIST}
+export SOME_K8S_VERSION=${SOME_K8S_VERSION}
+export TB_ORG=${TB_ORG}
+
 eval "$(grep '^ENV CATTLE_SYSTEM_AGENT' package/Dockerfile | awk '{print "export " $2 "=" $3}')"
+eval "$(grep '^ENV CATTLE_KDM_BRANCH' package/Dockerfile | awk '{print "export " $2 "=" $3}')"
+#export SOME_K8S_VERSION=$(curl https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/$CATTLE_KDM_BRANCH/data/data.json | jq -r ".$DIST.channels[0].latest")
```
[the remark this answers is not in the document]
Comment (Oats87) [reply scripts/provisioning-tests:42]: Comment added

File: scripts/provisioning-tests (hunk 4 above)
Comment (Oats87) [reply scripts/provisioning-tests:58]: Yes, I removed the `set -x`

File: .drone.yml (hunk 1 above)
Comment (Oats87) [reply .drone.yml:53]: I'm actually for this, so that we don't keep accidentally merging things that fail to build on publish.

It does slow down PR-build time, but the way the drone runners are set up, a drone runner is going to get provisioned for the specific pipeline even just doing a clone

File: go.mod [hunk 6]
```diff
@@ -105,8 +105,8 @@ require (
 	github.com/rancher/fleet/pkg/apis v0.0.0-20210918015053-5a141a6b18f0
 	github.com/rancher/gke-operator v1.1.2
 	github.com/rancher/kubernetes-provider-detector v0.1.5
-	github.com/rancher/lasso v0.0.0-20210709145333-6c6cd7fd6607
-	github.com/rancher/lasso/controller-runtime v0.0.0-20210608205930-775fcaf2f523
+	github.com/rancher/lasso v0.0.0-20220110205840-98715bdd6b5b
+	github.com/rancher/lasso/controller-runtime v0.0.0-20220110204826-724e3e57c47d
```
[the remark this answers is not in the document]
Comment (Oats87) [reply go.mod]: Removed

File: scripts/provisioning-tests [hunk 7]
```diff
@@ -43,9 +100,9 @@ while ! kubectl --kubeconfig /etc/rancher/k3s/k3s.yaml rollout status -w -n catt
 done
 #kill $TPID
 
-echo Running tests
+echo Running provisioning-tests
 export KUBECONFIG=/etc/rancher/k3s/k3s.yaml
-go test -v -timeout 40m ./tests/integration/pkg/tests/... || {
+go test -v -timeout 60m ./tests/integration/pkg/tests/custom/... || {
```
[the remark this answers is not in the document]
Comment (Oats87) [reply scripts/provisioning-tests]: lol, removed, thanks for catching that, I'm tired

File: tests/integration/pkg/tests/custom/custom_test.go [hunk 8]
```diff
@@ -32,6 +33,9 @@ func TestSystemAgentVersion(t *testing.T) {
 }
 
 func TestCustomOneNode(t *testing.T) {
+	if strings.ToLower(os.Getenv("DIST")) == "rke2" {
+		return
```
Comment (thedadams) [inline tests/integration/pkg/tests/custom/custom_test.go:37]: I think it's better to use `t.Skip()` for these to make it explicit. If we are actually going to do this.

File: .drone.yml (hunk 1 above)
Comment (rosskirkpat) [reply .drone.yml:53]: https://github.com/rancher/rancher/pull/35457

When I timed it last, it was around ten extra minutes per drone run to have both sets of provisioning tests run on PR.

**Review by rosskirkpat — APPROVED** [review]
LGTM

**Review by luthermonson — APPROVED** [review]

**Review by thedadams — APPROVED** [review]
